Adding support for the infinity point

bjk17 opened the pull request:
Every Elliptic Curve _E_ over a finite field contains the infinity point _O_ that satisfies
(a) _P_ + _O_ = _O_ + _P_ = _P_ for all points _P_ in _E_
(b) _P_ + (-_P_) = _O_ for all points _P_ in _E_

I added support for the infinity point into the library as the object `Point(None, None, curve)` and edited the code in a backwards-compatible way along with new tests. 

I appreciate your time reviewing this PR and if you end up accepting it then I suggest bumping the version to `1.1.0`.

bjk17 committed:
Added the infinity point and edited the addition algorithm to properly calculate with it

bjk17 commented:
Dear maintainer, 

I sincerely hope that you'll have time to take this PR into consideration. I've been using this library (with the changes in the PR) to solve examples on Elliptic Curve Cryptography from the book "An Introduction to Mathematical Cryptography" by J. Hoffstein, J. Pipher, J.H. Silverman. 

The book is used in a preliminary course on Cryptography at Reykjavik University and we want to be able to direct students to use this library to solve the exercises on ECC this fall semester. Without the proposed changes not all propblems can be solved, however.

Kavakuo commented:
Hi, thanks for your pull request. This went completely under my radar, sorry for not responding. I think it looks good so far. 
Maybe adding a property `isInfinityPoint` to the Point class is not the best solution. I thought about adding a static method to the Point class such as `Point.infinity()` returns the infinity point. To test if a point is the inifity point you can then do something like `p == Point.infinity()`.
What do you think about it?

bjk17 commented:
Hi! I'm not a fan of it, to be honest, but I guess it's down to syntactic preference. The strongest argument I have for `isInfinityPoint` property instead of comparing to `Point.infinity()` is that the latter needs to create a new object every time instead of just looking at variables. But I did change it as you suggested and I will push those changes to this PR if you prefer it that way. Have a nice day!

Kavakuo commented:
That's indeed a strong argument. The only "problem" I currently see is that `Point(None, None)` is very implicit. A solution for the object creation might be to return always the same object? Initialize the point of infinity at global scope and always return that one from the static method.

bjk17 commented:
> The only "problem" I currently see is that `Point(None, None)` is very implicit.

I agree. But I couldn't figure out a nicer way of denoting the infinity point. I'm very open for suggestions.

> Initialize the point of infinity at global scope and always return that one from the static method.

I don't think that is a good idea because it's a mutable object and the logic depends on it not changing.
```python
>>> from PyECCArithmetic import *
>>> o = Point.infinity()
>>> o.x = 42
>>> Point(None, None) == Point.infinity()
False
```

Kavakuo commented:
That's a good point as well. 
Maybe just provide both options? The code inside the class would get more explicit and easier to unterstand and the user could use whatever he likes most.

bjk17 committed:
Adding staticmethod Point.infinity() that returns a copy of the infinty point

bjk17 committed:
Caching the result wether point is the infinity point or not, in line with how isOnCurve() function is written

bjk17 commented:
I added the staticmethod `Point.infinity()` that returns a new Point object each time so there is no unexpected side effects of modifying that object.

Kavakuo committed:
Replaced Point(None, None) with static method

Kavakuo committed:
Added additional test for the infinity point

Kavakuo commented:
Pushed two additional commits as well. I would appreciate it, if you have time to take a look at it. If everything looks good to you, I would like to press the merge button.

bjk17 commented:
Hi! Your changes look fine. However, I now have some concerns about the static `Point.infinity()` method. I think it needs to take an optional `curve` parameter to set the correct curve. Two points cannot be added together if they're on different curves, it'll throw an `PointsOnDifferentCurveError`. I will comment on a specific line in the PR with an example in a moment.

bjk17 commented on tests.py:
```diff
@@ -23,6 +23,36 @@ def test_add(self):
 
+    def test_infinity(self):
+        P = self.results[0]
+        o = Point(None, None, self.curve)
```
If we'd change this line to `o = Point.infinity()` we'd get a `PointsOnDifferentCurveError` below when adding `P` and `o`.

Kavakuo committed:
Fix curve validation for calculations with PoI

Added a new method to check if two points are on the same curve. This basically fixes curve validation for calculations with the point of infinity.

bjk17 commented:
Amazing! :D I'm very happy with this now, merge if you agree :)

Kavakuo commented:
`IsInverseOf` also failed, thanks for the hint. It's not really a nice solution, but it works for now. I more and more get the feeling that the point of infinity should become an extra class, because `Point.infinity()` is currently assigned to the default curve, which is not really nice. Adding a new `curve` parameter to the method also doesn't feel right for me.

Anyway I merge this for now and publish a new version to PyPi so that you can use it in your course at university. Thanks for your help! I really appreciated it!

bjk17 commented:
Thank you very much. I really appreciate you taking time to solving the issue with me in order to get out a new version of the library with this feature that is needed for us here at the university. The EC part of the Crypto course is scheduled towards the end of the semester so don't be surprised if you get some feedback in November from a group of new users!

Kavakuo commented:
I would be really happy about that! If someone has other ideas how this tiny library can be improved, I'm very open for this. 
Last but not least, it would be great if you can tell your students about the [invalid curve attack](https://web-in-security.blogspot.com/2015/09/practical-invalid-curve-attacks.html), implementing this attack was my initial motivation to create this library 😃

Kavakuo commented:
Is published as [1.1.0 at PyPi](https://pypi.org/project/PyECCArithmetic/).
